aten/src/ATen/native/cuda/Bucketization.cu

```diff
@@ -0,0 +1,138 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/native/BucketizationUtils.h>
+#include <THC/THC.h>
+
+namespace at {
+namespace native {
+
+// Implement a TF like searchsorted and a bucketize function running on cuda
+// See details in ATen/nativate/Bucketization.cpp
+
+namespace {
+
+template<typename input_t>
+__device__ int64_t lower_bound(const input_t *data_ss, int64_t start, int64_t end, input_t val) {
+  while (start < end) {
+    int64_t mid = start + ((end - start) >> 1);
+    if (data_ss[mid] < val) {
+      start = mid + 1;
+    }
+    else {
+      end = mid;
+    }
+  }
+  return start;
+}
+
+template<typename input_t>
+__device__ int64_t upper_bound(const input_t *data_ss, int64_t start, int64_t end, input_t val) {
+  while (start < end) {
+    int64_t mid = start + ((end - start) >> 1);
+    if (data_ss[mid] <= val) {
```

Comment:
if val is nan, the position returned will be 0, which does not match numpy behavior (where for nan inputs the position is the last).

Reply:
Ah, I missed that. Will correct it.

Reply:
Also adding a test for nan inputs will be useful.

Reply:
sure